text/0000-extension-layer.md

```diff
@@ -0,0 +1,128 @@
+# Meta
+[meta]: #meta
+- Name: Add extension layer to exchange data
+- Start Date: 2023-10-09
+- Author(s): [c0d1ngm0nk3y](https://github.com/c0d1ngm0nk3y), [pbusko](https://github.com/pbusko)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request:
+- CNB Pull Request:
+- CNB Issue:
+- Related: [RFC#0105 Support Dockerfiles](https://github.com/buildpacks/rfcs/blob/main/text/0105-dockerfiles.md)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC introduces support for Extension configurable context to allow data transfer between the build environment and the Kaniko execution.
+
+# Motivation
+[motivation]: #motivation
+
+This change allows extensions to create their own context for the extend phase during the generation phase. Additionally, it ensures that extension output does not inadvertently interfere with other extension or buildpack layers during the build, and it does not unintentionally become part of the final application image.
+
+This would allow distroless run images to be extended.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This follows up on RFC-0105 and proposes that during the execution of the extension's `./bin/generate`, an extension is allowed to write arbitrary data to the `context` folder within its exclusive output directory. This data then becomes accessible during the execution of the `extend` phase via Kaniko build context. The content of these extension-specific context is ignored at build and launch time, it serves only the extension phase.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+- Before execution of the `./bin/generate`, the lifecycle will create a distinct writable layer `$CNB_GENERATED_DIR/<extension-id>` for each extension which passed detection.
+- The `$CNB_GENERATED_DIR/<extension-id>` is provided to the `./bin/generate` as `<output>` (`$CNB_OUTPUT_DIR`) directory.
```

Comment:
Currently the extender is looking for Dockerfiles in `<generated>/run/<image extension ID>/Dockerfile` and `<generated>/build/<image extension ID>/Dockerfile` (where `<generated>` defaults to `<layers>/generated`). Should we change the Platform API so that the extender looks for `<generated>/<image extension ID>/run.Dockerfile`, or should the lifecycle re-arrange the files after the generate phase? The former might be simpler in the long run, but platforms such as `pack` look for the presence of files in `<generated>/build` and `<generated>/run` to infer that extension is needed.

Reply:
I would go with writing to `<generated>/<image extension ID>/run.Dockerfile`  and/or `<generated>/<image extension ID>/build.Dockerfile`, as suggested in this RFC. Platforms such as `pack` might have a fallback mechanism, instead of adding more complex logic on the lifecycle side.

Reply:
@pbusko I'm in favor of removing this complexity from the lifecycle. Could we add an `extend` boolean to the `[build-image]` table in analyzed.toml, to match `[run-image]`:

```
[run-image]
  image = "<image>"
  reference = "<image reference>"
  extend = false

[build-image]
  reference = "<image reference>"
  extend = false
```

This will give platforms a simple way to check if extension is needed.